[ci] add benchmarks to pipeline

alvicsam opened the pull request:
Added benchmarks to pipeline

Part of https://github.com/paritytech/ci_cd/issues/174

alvicsam committed:
[WIP] add benchmarks to pipeline

alvicsam committed:
debug yarn run

alvicsam committed:
debug add timeout

alvicsam committed:
add lightweight-bench.lua

alvicsam committed:
run 5 tests in a row

alvicsam committed:
run benchmark 10 minutes

TarikGul commented on .gitlab-ci.yml:
```diff
@@ -96,6 +96,22 @@ build:
+    - yarn build
+    - yarn run main &
+    - sleep 10
+    - wrk -d10m -t4 -c12 --timeout 120s --latency -s ./ci/benchmarks/lightweight-bench.lua "http://127.0.0.1:8080"
```
This all seems incredibly flexible, thanks so much again! So if I understand this correctly we can actively update our benchmarking scripts through this directory, and fix any connecting calls through this yml, and it should proactively run the updated code?

alvicsam replied:
Yes, that's correct. I'm still experimenting and it seems that 10 minutes shows more accurate result.
This job runs on each commit. If we merge it with current settings it'll run on each commit in each pr and on master/tag as well.

alvicsam committed:
add variables

alvicsam committed:
move ci files; add artifacts

alvicsam committed:
change image

alvicsam committed:
add publish benchmark

alvicsam committed:
fix pipeline

alvicsam committed:
emulate specific metric

alvicsam committed:
finish

alvicsam committed:
fix rule

alvicsam committed:
fix rules

alvicsam commented:
@TarikGul 
I suggest merging this PR if it looks good for you.
Currently it's working like this:
- build and deploy logic is the same
- added jobs `benchmark`, `check-benchmark` and `push-benchmark`. They run directly after build and are not related to deploy jobs. They run only on master or on tag. 
- added `build-pr` and `benchmark-pr` jobs to test that docker image can be built during PR and run benchmarks manually if needed

`check-benchmark`  doesn't work because I need to adjust the script compares benchmark results. So while I'm doing this we can at least collect results

TriplEight commented on .gitlab-ci.yml:
```diff
@@ -96,20 +137,118 @@ build:
+  <<:                              *test-refs
+  script:
+    - echo building "$CONTAINER_REPO:$DOCKER_IMAGE_TAG"
+    - |-
+      if [[ $BUILD_LATEST_IMAGE ]]; then
+        buildah bud \
+        --format=docker \
+        --tag "$CONTAINER_REPO:$DOCKER_IMAGE_TAG" \
+        --tag "$CONTAINER_REPO:latest" "$DOCKERFILE_DIRECTORY"
+      else
+        buildah bud \
+        --format=docker \
+        --tag "$CONTAINER_REPO:$DOCKER_IMAGE_TAG" "$DOCKERFILE_DIRECTORY"
+      fi
```
without `|-` notation you could skip using `\` at the end of each line

TriplEight commented on .gitlab-ci.yml:
```diff
@@ -96,20 +137,118 @@ build:
+    PROMETHEUS_URL:                "http://vm-longterm.parity-build.parity.io"
+    CI_IMAGE:                      "paritytech/benchmarks:latest"
+  script:
+    - echo "TBD"
```
should it be in this PR?

alvicsam replied:
Nope, just made a stub job

TriplEight reviewed (changes requested):
minor grumbles

alvicsam committed:
fix fmt
